README.md

## Intro screen
simple splash screen with minimum configurations

- dynamic pages length
- configurable

**full example [here](https://pub.dev/packages/flutter_intro_screen/example)**

| page1                          | page2                          | page3                          |
| ------------------------------- | ------------------------------ | ------------------------------- |
| ![page1](readme/1.png) | ![page2](readme/2.png) | ![page3](readme/3.png) |


CHANGELOG.md
## 0.0.3

* TODO: fix on complete pressed

## 0.0.2

* TODO: update README.md
  
## 0.0.1

* TODO: Describe initial release.


test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_intro_screen/flutter_intro_screen.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    final widget = IntroScreen(
      appLogo: Row(
        children: [
          const Text("Splash"),
          Image.asset(
            "assets/images/intro-1.jpg",
            width: 20,
            height: 20,
          )
        ],
      ),
      pageData: [
        IntroPageData(
          title: const Text("World's First"),
          description: const Text("The world's first real estate app"),
          image: Image.asset(
            "assets/images/intro-1.jpg",
            fit: BoxFit.cover,
            alignment: Alignment.centerRight,
          ),
        ),
        IntroPageData(
          title: const Text("Google Map support"),
          description: const Text("Full support to iraq map"),
          image: Image.asset(
            "assets/images/intro-2.jpg",
            fit: BoxFit.cover,
            alignment: Alignment.centerRight,
          ),
        ),
      ],
      centerWidget: Container(),
      onCompletePressed: () {},
    );

    // Build our app and trigger a frame.
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: widget,
        ),
      ),
    );

    // Verify that our counter starts at 0.
    expect(find.text("World's First"), findsOneWidget);
    expect(find.text("Google Map support"), findsNothing);

    // // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byKey(const Key("World's First")));
    await tester.pump();

    // // Verify that our counter has incremented.
    // expect(find.text('0'), findsNothing);
    // expect(find.text('1'), findsOneWidget);
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_intro_screen/flutter_intro_screen.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    final widget = FlutterIntroScreen(
      appLogo: Row(
        children: [
          const Text("Splash"),
          Image.asset(
            "assets/images/intro-1.jpg",
            width: 20,
            height: 20,
          )
        ],
      ),
      pageData: [
        IntroPageData(
          title: const Text("World's First"),
          description: const Text("The world's first real estate app"),
          image: Image.asset(
            "assets/images/intro-1.jpg",
            fit: BoxFit.cover,
            alignment: Alignment.centerRight,
          ),
        ),
        IntroPageData(
          title: const Text("Google Map support"),
          description: const Text("Full support to iraq map"),
          image: Image.asset(
            "assets/images/intro-2.jpg",
            fit: BoxFit.cover,
            alignment: Alignment.centerRight,
          ),
        ),
      ],
      centerWidget: Container(),
      onCompletePressed: () {},
    );

    // Build our app and trigger a frame.
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: widget,
        ),
      ),
    );

    // Verify that our counter starts at 0.
    expect(find.text("World's First"), findsOneWidget);
    expect(find.text("Google Map support"), findsNothing);

    // // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byKey(const Key("World's First")));
    await tester.pump();

    // // Verify that our counter has incremented.
    // expect(find.text('0'), findsNothing);
    // expect(find.text('1'), findsOneWidget);
  });
}


lib/screen.dart
import 'package:extra_alignments/extra_alignments.dart';
import 'package:flutter/material.dart';
import 'package:flutter_animate/animate.dart';
import 'package:flutter_animate/effects/effects.dart';
import 'package:flutter_animate/num_duration_extensions.dart';
import 'package:gap/gap.dart';
import 'package:flutter_intro_screen/default_text_color.dart';
import 'package:flutter_intro_screen/indicator.dart';

class IntroDurations {
  final Duration fast;
  final Duration med;
  final Duration slow;
  final Duration pageTransition;

  const IntroDurations({
    this.fast = const Duration(milliseconds: 300),
    this.med = const Duration(milliseconds: 600),
    this.slow = const Duration(milliseconds: 900),
    this.pageTransition = const Duration(milliseconds: 200),
  });
}

class IntroMessages {
  final String swap;
  final String navigate;

  const IntroMessages({
    this.swap = "Swap left to continue",
    this.navigate = "Navigate",
  });
}

class IntroText {
  final TextStyle body;
  final TextStyle bodySmall;

  const IntroText({
    this.body = const TextStyle(),
    this.bodySmall = const TextStyle(),
  });
}

class IntroColors {
  final Color text;
  final Color button;
  final Color background;

  const IntroColors({
    this.background = Colors.black,
    this.text = Colors.white,
    this.button = Colors.white,
  });
}

class IntroInsets {
  final double xxs;
  final double xs;
  final double sm;
  final double md;
  final double lg;
  final double xl;
  final double xxl;

  const IntroInsets({
    this.xxs = 4,
    this.xs = 8,
    this.sm = 16,
    this.md = 24,
    this.lg = 32,
    this.xl = 48,
    this.xxl = 56,
  });
}

class FlutterIntroScreen extends StatefulWidget {
  final List<IntroPageData> pageData;
  final IntroMessages messages;
  final IntroText texts;
  final IntroInsets insets;
  final IntroDurations durations;
  final IntroColors colors;
  final VoidCallback onCompletePressed;

  final Widget appLogo;
  final Widget centerWidget;

  const FlutterIntroScreen({
    Key? key,
    required this.appLogo,
    required this.centerWidget,
    required this.pageData,
    required this.onCompletePressed,
    this.messages = const IntroMessages(),
    this.insets = const IntroInsets(),
    this.durations = const IntroDurations(),
    this.colors = const IntroColors(),
    this.texts = const IntroText(),
  }) : super(key: key);

  @override
  State<FlutterIntroScreen> createState() => _FlutterIntroScreenState();
}

class _FlutterIntroScreenState extends State<FlutterIntroScreen> {
  static const double _imageSize = 264;
  static const double _logoHeight = 126;
  static const double _textHeight = 155;
  static const double _pageIndicatorHeight = 55;

  late final PageController _pageController = PageController()
    ..addListener(_handlePageChanged);
  final ValueNotifier<int> _currentPage = ValueNotifier(0);

  @override
  void dispose() {
    _pageController.dispose();
    super.dispose();
  }

  void _handlePageChanged() {
    int newPage = _pageController.page?.round() ?? 0;
    _currentPage.value = newPage;
  }

  void _handleSemanticSwipe(int dir) {
    _pageController.animateToPage((_pageController.page ?? 0).round() + dir,
        duration: widget.durations.fast, curve: Curves.easeOut);
  }

  @override
  Widget build(BuildContext context) {
    // This view uses a full screen PageView to enable swipe navigation.
    // However, we only want the title / description to actually swipe,
    // so we stack a PageView with that content over top of all the other
    // content, and line up their layouts.

    final List<Widget> pages =
        widget.pageData.map((e) => _Page(data: e)).toList();

    final Widget content = Stack(children: [
      // page view with title & description:
      MergeSemantics(
        child: Semantics(
          onIncrease: () => _handleSemanticSwipe(1),
          onDecrease: () => _handleSemanticSwipe(-1),
          child: PageView(
            physics: const BouncingScrollPhysics(),
            controller: _pageController,
            children: pages,
            onPageChanged: (_) {},
          ),
        ),
      ),

      IgnorePointer(
        ignoringSemantics: false,
        child: Column(children: [
          const Spacer(),

          // logo:
          Semantics(
            header: true,
            child: Container(
              height: _logoHeight,
              alignment: Alignment.center,
              child: widget.appLogo,
            ),
          ),

          // masked image:
          SizedBox(
            height: _imageSize,
            width: _imageSize,
            child: ValueListenableBuilder<int>(
              valueListenable: _currentPage,
              builder: (_, value, __) {
                return AnimatedSwitcher(
                  duration: widget.durations.slow,
                  child: KeyedSubtree(
                    key: ValueKey(
                        value), // so AnimatedSwitcher sees it as a different child.
                    child: IntroPageImage(data: widget.pageData[value]),
                  ),
                );
              },
            ),
          ),

          // placeholder gap for text:
          const Gap(_FlutterIntroScreenState._textHeight),

          // page indicator:
          Container(
            height: _pageIndicatorHeight,
            alignment: const Alignment(0.0, -0.75),
            child: AppPageIndicator(
              count: widget.pageData.length,
              controller: _pageController,
              color: widget.colors.background,
            ),
          ),

          const Spacer(flex: 2),
        ]),
      ),

      // finish button:
      Positioned(
        right: widget.insets.lg,
        bottom: widget.insets.lg,
        child: _buildFinishBtn(context),
      ),

      // nav help text:
      BottomCenter(
        child: Padding(
          padding: EdgeInsets.only(bottom: widget.insets.lg),
          child: _buildNavText(context),
        ),
      ),
    ]);

    return DefaultTextColor(
      color: widget.colors.text,
      child: Container(
        color: widget.colors.background,
        child: SafeArea(child: content.animate().fadeIn(delay: 500.ms)),
      ),
    );
  }

  Widget _buildFinishBtn(BuildContext context) {
    return ValueListenableBuilder<int>(
      valueListenable: _currentPage,
      builder: (_, pageIndex, __) {
        return AnimatedOpacity(
          opacity: pageIndex == widget.pageData.length - 1 ? 1 : 0,
          duration: widget.durations.fast,
          child: FloatingActionButton.small(
            elevation: 0,
            onPressed: widget.onCompletePressed,
            tooltip: widget.messages.swap,
            backgroundColor: widget.colors.button,
            child: const Icon(Icons.arrow_forward_ios_rounded),
          ),
        );
      },
    );
  }

  Widget _buildNavText(BuildContext context) {
    return ValueListenableBuilder(
      valueListenable: _currentPage,
      builder: (_, pageIndex, __) {
        return AnimatedOpacity(
          opacity: pageIndex == widget.pageData.length - 1 ? 0 : 1,
          duration: widget.durations.fast,
          child: Semantics(
            onTapHint: widget.messages.swap,
            onTap: () {
              final int current = _pageController.page!.round();
              _pageController.animateToPage(
                current + 1,
                duration: 250.ms,
                curve: Curves.easeIn,
              );
            },
            child: widget.centerWidget,
          ),
        );
      },
    );
  }
}

@immutable
class IntroPageData {
  const IntroPageData({
    required this.title,
    required this.description,
    required this.image,
    this.mask,
  });

  final Widget title;
  final Widget description;
  final Widget image;
  final Widget? mask;
}

class _Page extends StatelessWidget {
  const _Page({Key? key, required this.data}) : super(key: key);

  final IntroPageData data;

  @override
  Widget build(BuildContext context) {
    return Semantics(
      liveRegion: true,
      child: Column(
        children: [
          const Spacer(),
          const Gap(_FlutterIntroScreenState._imageSize +
              _FlutterIntroScreenState._logoHeight),
          SizedBox(
            height: _FlutterIntroScreenState._textHeight,
            width: _FlutterIntroScreenState._imageSize,
            child: IntroStaticTextScale(
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  data.title,
                  const Gap(12),
                  data.description,
                ],
              ),
            ),
          ),
          const Gap(_FlutterIntroScreenState._pageIndicatorHeight),
          const Spacer(flex: 2),
        ],
      ),
    );
  }
}

class IntroPageImage extends StatelessWidget {
  const IntroPageImage({
    Key? key,
    required this.data,
  }) : super(key: key);

  final IntroPageData data;

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        SizedBox.expand(
          child: data.image,
        ),
        if (data.mask != null)
          Positioned.fill(
            child: data.mask!,
          ),
      ],
    );
  }
}

class IntroStaticTextScale extends StatelessWidget {
  const IntroStaticTextScale({
    Key? key,
    required this.child,
    this.scale = 1,
  }) : super(key: key);
  final Widget child;
  final double scale;

  @override
  Widget build(BuildContext context) {
    return MediaQuery(
      data: MediaQuery.of(context).copyWith(textScaleFactor: scale),
      child: child,
    );
  }
}


lib/indicator.dart
import 'package:flutter/material.dart';
import 'package:smooth_page_indicator/smooth_page_indicator.dart';

class AppPageIndicator extends StatefulWidget {
  const AppPageIndicator({
    Key? key,
    required this.count,
    required this.controller,
    required this.color,
    this.onDotPressed,
    this.dotSize,
    String? semanticPageTitle,
  })  : semanticPageTitle =
            semanticPageTitle ?? "strings.appPageDefaultTitlePage",
        super(key: key);
  final int count;
  final PageController controller;
  final void Function(int index)? onDotPressed;
  final Color color;
  final double? dotSize;
  final String semanticPageTitle;

  @override
  State<AppPageIndicator> createState() => _AppPageIndicatorState();
}

class _AppPageIndicatorState extends State<AppPageIndicator> {
  final _currentPage = ValueNotifier(0);

  @override
  void initState() {
    super.initState();
    widget.controller.addListener(_handlePageChanged);
  }

  int get _controllerPage => widget.controller.page?.round() ?? 0;

  void _handlePageChanged() => _currentPage.value = _controllerPage;

  @override
  Widget build(BuildContext context) {
    return Stack(children: [
      Container(
        color: Colors.transparent,
        height: 30,
        alignment: Alignment.center,
        child: ValueListenableBuilder<int>(
            valueListenable: _currentPage,
            builder: (_, value, child) {
              return Container();
            }),
      ),
      Positioned.fill(
        child: Center(
          child: ExcludeSemantics(
            child: SmoothPageIndicator(
              controller: widget.controller,
              count: widget.count,
              onDotClicked: widget.onDotPressed,
              effect: ExpandingDotsEffect(
                  dotWidth: widget.dotSize ?? 6,
                  dotHeight: widget.dotSize ?? 6,
                  paintStyle: PaintingStyle.fill,
                  strokeWidth: (widget.dotSize ?? 6) / 2,
                  dotColor: widget.color,
                  activeDotColor: widget.color,
                  expansionFactor: 2),
            ),
          ),
        ),
      ),
    ]);
  }
}


lib/default_text_color.dart
import 'package:flutter/material.dart';

class DefaultTextColor extends StatelessWidget {
  const DefaultTextColor({Key? key, required this.color, required this.child})
      : super(key: key);
  final Color color;
  final Widget child;

  @override
  Widget build(BuildContext context) {
    return DefaultTextStyle(
      style: DefaultTextStyle.of(context).style.copyWith(color: color),
      child: child,
    );
  }
}

class LightText extends StatelessWidget {
  const LightText({Key? key, required this.child}) : super(key: key);
  final Widget child;

  @override
  Widget build(BuildContext context) => DefaultTextColor(
        color: Colors.white,
        child: child,
      );
}

class DarkText extends StatelessWidget {
  const DarkText({Key? key, required this.child}) : super(key: key);
  final Widget child;

  @override
  Widget build(BuildContext context) => DefaultTextColor(
        color: Colors.black,
        child: child,
      );
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:google_fonts/google_fonts.dart';
import 'package:flutter_intro_screen/flutter_intro_screen.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    // read from shared preference or data base
    bool showIntroScreen = true;

    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: showIntroScreen ? const IntroScreen() : const MyHomePage(),
    );
  }
}

class IntroScreen extends StatefulWidget {
  const IntroScreen({Key? key}) : super(key: key);

  @override
  State<IntroScreen> createState() => _IntroScreenState();
}

class _IntroScreenState extends State<IntroScreen> {
  @override
  Widget build(BuildContext context) {
    const colors = IntroColors(
      button: Color(0xfff39c12),
      text: Colors.white,
      background: Color(0xff0d3352),
    );

    return Scaffold(
      body: FlutterIntroScreen(
        onCompletePressed: () {
          Navigator.of(context).push(
            MaterialPageRoute(
              builder: (context) {
                return const MyHomePage();
              },
            ),
          );
        },
        colors: colors,
        appLogo: _buildAppLogo(),
        pageData: _buildPageData(),
        centerWidget: Text(
          "swap left to continue",
          style: GoogleFonts.raleway(),
        ),
      ),
    );
  }

  Widget _buildAppLogo() {
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: [
        const Icon(
          Icons.house,
          color: Colors.white,
        ),
        const SizedBox(height: 12),
        Text(
          "Example App",
          style: GoogleFonts.yesevaOne(fontSize: 24),
        ),
      ],
    );
  }

  List<IntroPageData> _buildPageData() {
    final titleStyle = GoogleFonts.yesevaOne(fontSize: 24);
    final descriptionStyle = GoogleFonts.raleway(fontSize: 16);
    return <IntroPageData>[
      IntroPageData(
        title: Text(
          "Find it",
          style: titleStyle,
        ),
        description: Text(
          "filters, maps, and search",
          textAlign: TextAlign.center,
          style: descriptionStyle,
        ),
        image: const IntroAssetsImage(path: "assets/images/intro-1.jpg"),
      ),
      IntroPageData(
        title: Text(
          "Tour it.",
          style: titleStyle,
        ),
        description: Text(
          "read all the details about the property",
          textAlign: TextAlign.center,
          style: descriptionStyle,
        ),
        image: const IntroAssetsImage(path: "assets/images/intro-2.jpg"),
      ),
      IntroPageData(
        title: Text(
          "Own it.",
          style: titleStyle,
        ),
        description: Text(
          "Direct contact with the owner of the property",
          textAlign: TextAlign.center,
          style: descriptionStyle,
        ),
        image: const IntroAssetsImage(path: "assets/images/intro-3.jpg"),
      ),
    ];
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: const Center(
        child: Text('Flutter Demo Home Page'),
      ),
    );
  }
}
